s3store/s3store.go

```diff
@@ -547,3 +561,28 @@ func isAwsError(err error, code string) bool {
 	}
 	return false
 }
+
+func (store S3Store) CalcOptimalPartSize(size *int64) int64 {
+	switch {
+	// We can only manage files up to MaxObjectSize, else we need to fail.
+	case *size > store.MaxObjectSize:
+		return 0
+	// When upload is smaller or equal MinPartSize, we upload in just one part.
+	case *size <= store.MinPartSize:
+		return store.MinPartSize
+	// When we need 9999 parts or less with MinPartSize.
+	case *size/store.MinPartSize < store.MaxMultipartParts:
+		return store.MinPartSize
+	// When we can divide our upload into exactly MaxMultipartParts parts with
+	// no bytes leftover, we will not need an spare last part.
+	// Also, when MaxObjectSize is equal to MaxPartSize * MaxMultipartParts
+	// (which is not the case with the current AWS S3 API specification, but
+	// might be in the future or with other S3-aware stores), we need this in
+	// order for our Multipart-Upload to reach full MaxObjectSize.
+	case int64(math.Mod(float64(*size), float64(store.MaxMultipartParts))) == 0:
+		return *size / store.MaxMultipartParts
+	// In all other cases, we need a spare last piece for the remaining bytes.
+	default:
+		return *size / (store.MaxMultipartParts - 1)
```

Comment:
I am afraid, you misread `MaxMultipartParts` here. The "- 1" in `(store.MaxMultipartParts - 1) `does not reduce the part size, but the total number of parts, we are dividing up our upload by. The part size actually grows due to that.

- MaxMultipartParts
is the maximum number of parts an S3 multipart upload is allowed to have according to AWS S3 API specifications.

However, compiling this answer, I actually found a bug in the algorithm, another special case, which needs to be checked for. I'll attend to that at the bottom of this post.

As for math.Mod, I just realised, there is the "%" operator for the remainder of an integer division, so of course that one is better suited, as it produces predictable results with no rounding problems. I will switch over to that.

As for the algorithm per se, the division by `MaxMultipartParts` will almost never produce an exact partition, but a number rounded down to an integer and there will be a remainder to this division. That remainder is what will end up in the last spare part. As we expect there to be a remainder, we need to divide the upload size by 9999 not 10.000, so we have a part left for the remainder. That is what is going on here.

The section above the `default:` checks, whether a division by 10.000 is applicable, by checking if a division `size / MaxMultipartParts` will produce any remainder or not. And only if it is a division without a remainder, we divide by `MaxMultipartParts` not `(MaxMultipartParts-1)` (10.000 not 9.999 in our case).

Here are a few examples with simple numbers for illustration:
MaxMultipartParts = 10
MinPartSize = 5
MaxPartSize = 10
MaxObjectSize = 100

size = 100
size / MaxMultipartParts  
= 100 / 10 = 10 with no remainder 
... No remainder, so we will divide by MaxMultipartParts and have 10 parts with size 10.
Total: 10 parts - SUCCESS.

size = 98
size / MaxMultipartParts  
= 99 / 10 = 9 with a remainder of 8
... So if we would use the result of this division as our optimalPartSize, then we would end up with 10 parts of size 9 and an 11th part of size 8. Which does not work, as we would exceed MaxMultipartParts.

You could say, well then let's just use the integer result of this division as the optimalPartSize (ergo 9). The fact that this division seemingly already produces, what we expect as optimalPartSize, is only a result of the fact, that MaxMultipartParts and MaxPartSize both are 10 in our example. So don't be mislead and rather ignore that fact.

Next we divide the size by (MaxMultipartParts - 1), so we have one spare part left, just in case.
size = 98
size / (MaxMultipartParts - 1)
= 99 / 9 = 10 with a remainder of 8
... So we have 9 parts of size 10 and a 10th part of size 8. 
Total: 10 parts - SUCCESS

I skipped size = 99, as it is a special case, and actually reveals a bug in the current algorithm:
size = 99
size / MaxMultipartParts
= 99 / 10 = 9 with a remainder of 9
... That means, we would end up with 10 parts of size 9 and an 11th part of size 9 again. In the current algorithm, as we find a rest in this division, we need to progress to the next case. 

But if we divide by (MaxMultipartParts - 1), we wont be lucky as well in this case:
= 99 / 9 = 11 with no remainder in that special case
... we end up with 9 parts of size 11, which exceeds MaxPartSize.

So we need to add another check:
case: size % (MaxMultiparts - 1) > MaxPartSize 
(ergo, if the last part would be larger than MaxPartSize)
then optimalPartSize = MaxPartSize
... we end up with 9 parts of size 10 and a last part of size 9. 
Total: 10 parts - SUCCESS

I'll play that with a few other numbers:
MaxMultipartParts = 20
MinPartSize = 5
MaxPartSize = 10
MaxObjectSize = 200

size 200
200 / 20 = 10 no remainder
... 20 parts with size 10. 
Total: 20 parts - SUCCESS.

size 199
199 / 20 = 9 with a remainder of 19
199 / 19 = 10 with a remainder of 9. The remainder is less than MaxPartSize, so apply the default.
... 19 parts with size 10 and a last one with size 9. 
Total: 20 parts - SUCCESS.

size 180
180 / 20 = 9 with no remainder
... 20 parts with size 9. 
Total: 20 parts - SUCCESS.

size 181
181 / 20 = 9 with a remainder of 1
181 / 19 = 9 with a remainder of 10. The remainder is not greater than MaxPartSize, so we apply default.
... 19 parts with size 9 and a last one with size 10. 
Total: 20 parts - SUCCESS.

We could also apply optimalPartSize = MaxPartSize here, so we would end up with 18 parts with size 10 and a last one with size 1. Total: 19 parts. As we thrive for the maximum number of parts with a minimum PartSize, the former solution is better.

size 182
182 / 20 = 9 with a remainder of 2
182 / 19 = 9 with a remainder of 11. The remainder is greater than MaxPartSize, so we apply MaxPartSize.
... 18 parts with size 10 and a last one with size 2. 
Total: 19 parts - SUCCESS.

size 189
189 / 20 = 9 with a remainder of 9
189 / 19 = 9 with a remainder of 18. The remainder is greater than MaxPartSize, so we apply MaxPartSize.
... 18 parts with size 10 and a last one with size 9. 
Total: 19 parts - SUCCESS.

size 190
190 / 20 = 9 with a remainder of 10 remainder
190 / 19 = 10 with no remainder. The remainder is less than MaxPartSize, so we apply default.
... 19 parts with size 10. 
Total: 19 parts - SUCCESS.

size 191
191 / 20 = 9 with a remainder of 11. We have a remainder, so we proceed to next case.
191 / 19 = 10 with a remainder of 1. The remainder is less than MaxPartSize, so we apply default.
... 19 parts with size 10 and a last one with size 1. 
Total: 20 parts - SUCCESS.

The addition of this additional switch should produce a universally applicable algorithm and replacing math.Mod with "%" operator, will fix your other concern.